File: 0003-best-time-to-buy-and-sell-stock/solution.go
func maxProfit(prices []int) int {
	diff := 0

	for i := 0; i < len(prices)-1; i++ {

		for j := i + 1; j < len(prices); j++ {
			if prices[j]-prices[i] > diff {
				diff = prices[j] - prices[i]
			}
		}
	}
	return diff
}

func maxProfitOptimized(prices []int) int {
	minPrice := math.MaxInt
	maxProfit := 0

	for _, price := range prices {
		if price < minPrice {
			minPrice = price
		} else if price-minPrice > maxProfit {
			maxProfit = price - minPrice
		}
	}
	return maxProfit
}

File: 0003-best-time-to-buy-and-sell-stock/solution.swift

class Solution {
    func maxProfit(_ prices: [Int]) -> Int {
        let input = [7,1,5,3,6,4]
        var diff = 0
        for i in 0..<input.count {
            for j in i+1..<input.count {
                let currentDiff = input[j] - input[i]
                if currentDiff > diff {
                    diff = currentDiff
                }
            }
        }
        return diff
    }
//    Time Complexity: O(n²) || Space Complexity: O(1)
    
    func maxProfitOptimized(_ prices: [Int]) -> Int {
        var minPrice = prices[0]
        var maxProfit = 0
        
        for price in prices {
            if price < minPrice {
                minPrice = price
            } else {
                let profit = price - minPrice
                if profit > maxProfit {
                    maxProfit = profit
                }
            }
        }
        return maxProfit
    }
}
//Time: n × O(1) = O(n) || Space: O(1)
